src/interface/sparse_pack.hpp

```diff
@@ -0,0 +1,282 @@
+  }
+  KOKKOS_INLINE_FUNCTION
+  static bool regex() { return REGEX; }
+  KOKKOS_INLINE_FUNCTION
+  static int ndim() { return sizeof...(NCOMP); }
+  KOKKOS_INLINE_FUNCTION
+  static int size() { return multiply<NCOMP...>::value; }
+
+  const int idx;
+};
+
+struct any : public base_t<true> {
+  template <class... Ts>
+  KOKKOS_INLINE_FUNCTION any(Ts &&...args)
+      : parthenon::variables::base_t<true>(std::forward<Ts>(args)...) {}
+  static std::string name() { return ".*"; }
+};
+} // namespace variables
+
+template <class... Ts>
+class SparsePack : public SparsePackBase {
+ public:
+  SparsePack() = default;
+
+  explicit SparsePack(const SparsePackBase &spb) : SparsePackBase(spb) {}
+
+  template <class T>
+  static SparsePack Make(T *pmd, const std::vector<MetadataFlag> &flags = {},
+                         bool fluxes = false, bool coarse = false) {
+    auto &cache = pmd->GetSparsePackCache();
+    return SparsePack(cache.Get(
+        pmd, PackDescriptor(std::vector<std::string>{Ts::name()...},
+                            std::vector<bool>{Ts::regex()...}, flags, fluxes, coarse)));
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  Make(T *pmd, const VAR_VEC &vars, const std::vector<MetadataFlag> &flags = {},
+       bool fluxes = false, bool coarse = false) {
+    static_assert(sizeof...(Ts) == 0);
+    auto out = SparsePackBase::Make(pmd, vars, flags, fluxes, coarse);
+    return {SparsePack(std::get<0>(out)), std::get<1>(out)};
+  }
+
+  template <class T>
+  static SparsePack MakeWithFluxes(T *pmd, const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = false;
+    const bool fluxes = true;
+    return Make(pmd, flags, fluxes, coarse);
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  MakeWithFluxes(T *pmd, const VAR_VEC &vars,
+                 const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = false;
+    const bool fluxes = true;
+    Make(pmd, vars, flags, fluxes, coarse);
+  }
+
+  template <class T>
+  static SparsePack MakeWithCoarse(T *pmd, const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = true;
+    const bool fluxes = false;
+    return Make(pmd, flags, fluxes, coarse);
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  MakeWithCoarse(T *pmd, const VAR_VEC &vars,
+                 const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = true;
+    const bool fluxes = false;
+    Make(pmd, vars, flags, fluxes, coarse);
+  }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNBlocks() const { return nblocks_; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNDim() const { return ndim_; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetDim(const int i) const {
+    assert(i > 0 && i < 6);
+    PARTHENON_REQUIRE(
+        i != 2, "Should not ask for the second dimension since it is logically ragged");
+    return dims_[i];
```

Comment:
Thinking ahead, if faces and cells and edges are all mixed in a pack, perhaps we should only allow one to ask for `GetDim(5)` and `GetDim(4)`.

My thinking here is that we always know the dim in the k, j, i dimensions. All we need to know are the numbers of blocks and the numbers of variables.

Reply:
Agreed, this was a design obstacle throughout the existing variable packs that I was struggling to design around - each different face and edge will have a different dimension from cell or node centered variables.

Reply:
I think that makes sense, I was just trying to partially adhere to the pre-existing pack interface. I will remove this.

Reply:
:+1: I like perhaps just a method for `GetNumBlocks` and your per-variable dimension methods that you already have.